### ISSUE TITLE
bug: Finalizer added to an object by a controller running on a core is not scrubbed before downsync to the wec

### ISSUE THREAD

**Issue by cwiklik** [kubestellar/kubestellar#1864]
### Describe the bug

The mcad application has multiple controllers, one running on the core (dispatcher) and another (agent) running on each of the wec's. When appwrapper CR object is created on the core, the dispatcher controller adds a finalizer to it. When this appwrapper is pushed downstream to the wec, the finalizer is not scrubbed (as part of downsync) preventing the object from being deleted. This issue is related to [#1851](https://github.com/kubestellar/kubestellar/issues/1851).

### Steps To Reproduce

See above

### Expected Behavior

Finalizers should be scrubbed/stripped from an object before downsync to the remote wec. The finalizer is created/owned by a controller on the core and should not be carried over to the wec since no controller there would be aware of it.  

### Additional Context

The current workaround for this issue is to make mcad agent controller aware of the dispatcher's finalizer and delete it as necessary.

**Comment by nirrozenbaum** [comment]
@cwiklik can you add more details?
which version of KubeStellar are you using?
can you please put the AW yaml you're deploying + what you see on the WEC?

**Comment by cwiklik** [comment]
I am running with Kubestellar version 0.20.0. For some reason I am not able to use attachment for the yaml so my test appwrapper is: 

apiVersion: workload.codeflare.dev/v1beta1 # new apiVersion
kind: AppWrapper
metadata:
  name: appwrapper-sample-job
  labels:
     appwrapper.mcad.ibm.com/assignedCluster: wec3
spec:
  priority: 5
  schedulingSpec:
    minAvailable: 1
  resources:
    GenericItems:
    - replicas: 1
      completionstatus: Complete
      custompodresources:
      - requests:
          memory: 10Mi
          cpu: 10m
        replicas: 1
      generictemplate:
        apiVersion: batch/v1
        kind: Job
        metadata:
          namespace: default
          name: sample-job
        spec:
          parallelism: 1
          completions: 1
          template:
            metadata:
              labels:
                appwrapper.mcad.ibm.com: placeholder # pod spec label
            spec:
              restartPolicy: Never
              containers:
                - name: busybox
                  image: quay.io/project-codeflare/busybox
                  imagePullPolicy: IfNotPresent
                  command: ["sh", "-c", "sleep 10"]
                  resources:
                    requests:
                      memory: 10Mi
                      cpu: 10

On the WEC, the relevant part of this appwrapper is:

apiVersion: workload.codeflare.dev/v1beta1
kind: AppWrapper
metadata:
  annotations:
    appwrapper.mcad.ibm.com/serializedStatus: 
  creationTimestamp: "2024-03-07T15:53:46Z"
  finalizers:
  - workload.codeflare.dev/finalizer_dispatcher       <<<<<<<<<<<< Created by mcad dispatcher
  - workload.codeflare.dev/finalizer_runner

**Comment by nirrozenbaum** [comment]
@cwiklik something is not clear.
the AW you deployed to KubeStellar WDS doesn't have any finalizer.
can you explain the part of `Created by MCAD dispathcer`? 
what is exactly the flow?
 is it the following?
 you apply AW to WDS > MCAD dispatcher adds finalizer to AW on the hub -> AW is distributed to WEC with finalizer?
 OR
 you apply AW to WDS > AW distributed to WEC -> MCAD dispatcher adds finalizer to AW on the WEC?

**Comment by cwiklik** [comment]
Let me try again, in this scenario Kubestellar (v0.20.0) is configured with WDS of type host. The Flow:

1) Apply AW on the host (WDS)
2) MCAD dispatcher controller (running on the host) adds a finalizer to the AW
3) MCAD dispatcher controller creates BindingPolicy for the AW
4) Kubestellar syncs down the AW to the WEC
5) AW on the WEC is picked up by MCAD agent controller and the workload is deployed and completes
6) The Status propagates from the WEC to the host
7) Using kubectl I delete AW
8) The AW lingers on the WEC since it contains a finalizer created by the dispatcher in step 2 (above)

**Comment by nirrozenbaum** [comment]
ok. now the flow is clear to me.
If I'm not mistaken I think this should be solved in v0.21.0 which should be released in the next few days.
would it be possible to re-test the same in few days using the new version? (after we cut the release)

**Comment by cwiklik** [comment]
Sorry for not being clear earlier. I think there are many words describing the same thing. Core, Hub, Host, WDS (of type host) mean the same thing (at least to me).  

When there is a new release I can easily verify the fix.

**Comment by cwiklik** [comment]
@nirrozenbaum I will be testing today 0.21.0 today and will let you know.

**Comment by cwiklik** [comment]
@nirrozenbaum @pdettori Your fix for removing finalizers created by a controller on the host works. The finalizer I had a problem with is no longer propagating to the WEC. Thanks.

**Comment by nirrozenbaum** [comment]
@cwiklik perfect, that's very good :)
do you think we can close this issue as fixed? or anything else we need to verify before closing?

**Comment by cwiklik** [comment]
@nirrozenbaum Yes, this issue can be closed since I fixes the problem.

**Comment by nirrozenbaum** [comment]
closed as fixed
